### PR TITLE
RUN-950: Changing default SCM SSH authentication to SSHJ

### PR DESCRIPTION
**Pull request by ltamaster**
**Is this a bugfix, or an enhancement? Please describe.**
enhancement:
Replacing default SCM SSH authentication (using JSCH) by SSHJ

**Describe the solution you've implemented**

**Describe alternatives you've considered**

**Additional context**
GitHub has tightened up their SSH key requirements, and wants SSH-RSA 2 keys. The new default for ssh-keygen via OpenSSH is not supported by JSCH, and is likely why we saw an influx of issues involving the error "invalid privatekey"

### CODE REVIEW

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSessionFactory.groovy [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        if (config == null)
+            config = OpenSshConfig.get(fs)
+
+        final OpenSshConfig.Host hc = config.lookup(host)
+        if (port <= 0)
+            port = hc.getPort()
+        if (user == null)
+            user = hc.getUser()
+
+        sshClient = createConnection(user, host,port)
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSessionFactory.groovy:55]: Is there any specific reason to make `sshClient` a property of the factory instance? The property value is overwritten every time `getSession` gets called, which means the old one got discarded.

A session factory is better to be an immutable singleton object without saving a SSHClient as its internal state.

According to the document about SSHClient and RemoteSession, they seem to be an abstraction of the same thing - an established SSH tunnel connected to a host. I think it's natural that an SSHClient instance should be part of the SshjSession instance.  

```suggestion
        return new SshjSession(createConnection(user, host, port)
```

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSessionFactory.groovy [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        return new SshjSession(sshClient, uri)
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSessionFactory.groovy:56]:
```suggestion
        
```

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSessionFactory.groovy [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            ssh.loadKnownHosts()
+        }else{
+            ssh.addHostKeyVerifier(new PromiscuousVerifier())
+        }
+
+        if (port != null) {
+            ssh.connect(host, port)
+        } else {
+            ssh.connect(host);
+        }
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSessionFactory.groovy:82]: SSHClient is an auto-closable resource. It's better to be enclosed in a try-with-resources block.
```
try(ssh.connect(host, port)) {
}
```

**Comment by carlosrfranco** [comment]
Tested using Github and it worked

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        this.uri = uri
+
+        this.sshClient = createConnection()
+
+    }
+
+    @Override
+    Process exec(String commandName, int timeout) throws IOException {
+        this.session = sshClient.startSession()
+        return new SshjProcess(commandName, timeout)
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy:39]: To utilize the autoclosable feature of Session and Command:
```
try(Session session = sshClient.startSession()
    Session.Command command = session.exec(commandName)
) {
   return new SshProcess(command, timeout)
}
```

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import org.eclipse.jgit.errors.TransportException
+import org.eclipse.jgit.transport.OpenSshConfig
+import org.eclipse.jgit.transport.RemoteSession
+import org.eclipse.jgit.transport.URIish
+import org.eclipse.jgit.util.io.IsolatedOutputStream
+
+@CompileStatic
+class SshjSession implements RemoteSession {
+
+    Session session
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy:19]: This property can be removed.

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    private class SshjProcess extends Process {
+
+        int timeout
+        Session.Command cmd
+        InputStream inputStream
+        OutputStream outputStream
+        InputStream errorStream
+
+        SshjProcess(String commandName, int tms) throws TransportException, IOException {
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy:99]:
```
SshjProcess(Session.Command command, int tms) throws TransportException, IOException {
```

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+            try {
+                timeout = tms
+                cmd = session.exec(commandName)
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy:102]: cmd = command

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                throw new IllegalStateException()
+            }
+            return cmd.getExitStatus()
+        }
+
+        @Override
+        void destroy() {
+            cmd.close()
+            closeOutputStream()
+            session.close()
```
Comment (ehe-pd) [inline plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy:134]: Can be changed to:
```suggestion: 
         if(cmd != null) cmd.close() 
         closeOutputStream()
```

**Review by ehe-pd — CHANGES_REQUESTED** [review]
The refactor looks good. 
There are a few improvements I think are worth applying.

File: plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy (hunk 4 above)
Comment (ltamaster) [reply plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSession.groovy:39]: I tried this but I got an error: Stream is closed, so maybe we can leave it as it is now

**Review by ehe-pd — APPROVED** [review]

**Pull request merged**